homeassistant/components/fitbit/application_credentials.py

```diff
@@ -0,0 +1,77 @@
+"""application_credentials platform the fitbit integration.
+
+See https://dev.fitbit.com/build/reference/web-api/authorization/ for additional
+details on Fitbit authorization.
+"""
+
+import base64
+import logging
+from typing import Any, cast
+
+from homeassistant.components.application_credentials import (
+    AuthImplementation,
+    AuthorizationServer,
+    ClientCredential,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import OAUTH2_AUTHORIZE, OAUTH2_TOKEN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FitbitOAuth2Implementation(AuthImplementation):
+    """Local OAuth2 implementation for Fitbit.
+
+    This implementation is needed to send the client id and secret as a Basic
+    Authorization header.
+    """
+
+    async def async_resolve_external_data(self, external_data: dict[str, Any]) -> dict:
+        """Resolve the authorization code to tokens."""
+        session = async_get_clientsession(self.hass)
+        data = {
+            "grant_type": "authorization_code",
+            "code": external_data["code"],
+            "redirect_uri": external_data["state"]["redirect_uri"],
+        }
+        resp = await session.post(self.token_url, data=data, headers=self._headers)
+        resp.raise_for_status()
+        return cast(dict, await resp.json())
+
+    async def _token_request(self, data: dict) -> dict:
+        """Make a token request."""
+        session = async_get_clientsession(self.hass)
+
+        data["client_id"] = self.client_id
+        if self.client_secret is not None:
```

Comment:
I've updated this to not need the `is None` check since it should always be present. This was code copied from the parent class and modified slightly, but i see its simpler to just reimplement how it makes sense here. I've also added a more detailed test for what is happening here with the refresh body.

I think this gets at the spirit of the comment.